File: singletonDP/cpp/singleton.cpp
#include <iostream>
#include <thread>
#include "logger.hpp"
using namespace std;

void user1(){
    Logger *logger1 = Logger::getLogger();
    logger1->Log("This msg is from user1");
}

void user2(){
    Logger *logger2 = Logger::getLogger();
    logger2->Log("This msg is from user2");
}

int main(){
    thread t1(user1);
    thread t2(user2);
    t1.join();
    t2.join();
    return 0;
}

//command to execute for thread safe in Linux --> g++ -o singletonDP *.cpp -pthread

//without thread --> g++ -o singletonDP *.cpp 

File: singletonDP/singletonPackage/singleton.java
package singletonPackage;

public class singleton extends Thread {
    public static void main(String[] args) {
        singleton thread1 = new singleton();
        singleton thread2 = new singleton();
        thread1.user1();
        thread2.user2();
    }

    void user1(){
        logger logger1 = logger.getLogger();
        logger1.Log("Logger1 logs");
    }

    void user2(){
        logger logger2 = logger.getLogger();
        logger2.Log("Logger2 logs");
    }
}
